Prevent crashes during and after cleanup

msmolens opened the pull request:
This commit prevents crashes by handling scenarios such as:
(a) object destruction after the Python interpreter has been finalized
(b) object destruction after cleanup, i.e. the singleton no longer exists

Any usage of a Qt enum demonstrates (a).

One example that demonstrates (b) is a QTimer object which is created with a
QApplication parent. PythonQt::cleanup() is called before the QApplication is
completely destroyed, so the code that handles wrapping the QTimer object must
handle the case when the PythonQt singleton no longer exists.

Based on https://github.com/commontk/PythonQt/commit/e7860427ddb0da2ff269d46431217aa02ccd1a90

jcfr commented:
Looks good. 

In my original commit  e786042, is the part about `removeSignalEmitter` needed ?

May you could add  `Co-authored-by:  Jean-Christophe Fillion-Robin <jchris.fillionr@kitware.com>` :) 

Do you think you could add a test crashing without the fix, and passing with the fix. That would basically be an other commit

patmarion commented:
If it's helpful, here's a pure C++ example that demonstrated the enum crash: https://gist.github.com/patmarion/4bf2905e5f8b3c0b09d2

jcfr commented:
@patmarion I confirm that the fix integrated by @msmolens fixes the issue (at least on ubuntu 14.04.03) :8ball:  Thanks for the gist :+1: 

@msmolens  Here is a topic where I added @patmarion test case. Feel free to rename the test of even integrate in the existing PythonQt test suite if that makes sense ...

msmolens committed:
Prevent crashes during and after cleanup

This commit prevents crashes by handling scenarios such as:
(a) object destruction after the Python interpreter has been finalized
(b) object destruction after cleanup, i.e. the singleton no longer exists

Any usage of a Qt enum demonstrates (a).

One example that demonstrates (b) is a QTimer object which is created with a
QApplication parent. PythonQt::cleanup() is called before the QApplication is
completely destroyed, so the code that handles wrapping the QTimer object must
handle the case when the PythonQt singleton no longer exists.

Co-authored-by: Jean-Christophe Fillion-Robin <jchris.fillionr@kitware.com>

msmolens committed:
Fix tests link error on Windows

This commit fixes a link error building PythonQtCppTests on Windows:

    1>PythonQtCppTests.obj : error LNK2001: unresolved external symbol "int __cdecl tests_PythonQtTestMain(int,char * * const)" (?tests_PythonQtTestMain@@YAHHQEAPEAD@Z)
    1>C:\temp\PythonQt-build\Debug\PythonQtCppTests.exe : fatal error LNK1120: 1 unresolved externals

msmolens committed:
Fix refcount problems seen when re-initializing Python after finalizing

A sequence of calls like the following would crash in Python when reinitializing
the interpreter the second time:

    PythonQt::init(0);
    ...
    Py_Finalize();

    PythonQt::init(0);
    ...
    Py_Finalize();

msmolens committed:
Explicitly initialize global storage containers

This commit changes initialization of the global storage containers to be
explicit instead of happening during static initialization. This makes the
containers properly initialized if PythonQt is initialized and cleaned up more
than once.

The motivation for this change is to support testing cleanup and finalization.

msmolens committed:
Add cleanup/finalization tests

Add tests that check for clean cleanup and finalization in different scenarios.

Co-authored-by: Jean-Christophe Fillion-Robin <jchris.fillionr@kitware.com>
Co-authored-by: Pat Marion <james.patrick.marion@gmail.com>

msmolens commented:
@jcfr:
I couldn't find a testcase that required the `removeSignalEmitter` change, but since it makes the the code more robust, and it was a problem you hit earlier, I've integrated the change.

I added a test similar to what @patmarion provided, but integrated it into the existing test suite. This allows having more than one test that exercises the PythonQt::init(...) ... Py_Finalize() workflow in a single project/executable. I added another test that runs a similar workflow as the Slicer test mentioned in https://github.com/Slicer/Slicer/pull/334#issuecomment-139264541.

I hit some more problems in PythonQt while setting this up; fixes are included.

jcfr commented:
:8ball:
